### PR TITLE
You can exclude already discounted products from voucher

### PR DESCRIPTION
**Pull request by kpodemski**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | This is rework of #2060 |
| Type? | improvement |
| Category? | BO |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? | many feature requests i guess ;) |
| How to test? | Add new voucher with selected option "Exclude specials" to "Yes", try to use voucher while having already discounted products in cart, if you have more than one product and one product is not discounted voucher will work only for this product |
| Sponsor company | impSolutions

Vouchers didn't work at all before this fix for undefined notice, i think it will be good to add tests to StarterTheme spec for this mechanism!

### CODE REVIEW

**Comment by vincentbz** [comment]
Thanks ! I try to test it today :+1:

**Comment by Shudrum** [comment]
Hi @kpodemski,

Sorry to ask this, but can you rebase your branch on the top of develop so we can test it easily?

Thank you!

**Comment by kpodemski** [comment]
Hi @Shudrum 

Done

**Comment by kpodemski** [comment]
@vincentbz news?

**Comment by vincentbz** [comment]
i'm on it now :)

**Comment by vincentbz** [comment]
I have this error when i save the voucher in creation : 

image

**Comment by vincentbz** [comment]
quick question : do I have to reinstall the shop before testing ?

**Comment by kpodemski** [comment]
Ni, just use code for alter table from upgrade sql, it will be faster, remember to change prefix 😃

**Comment by vincentbz** [comment]
It's better like this yes :)
Why the exclusion of "products on special" is only shown when the discount is based on % and not amount ?

**Comment by kpodemski** [comment]
This can be done later, for now it is only rework of older PR ad few bugfixes

**Comment by vincentbz** [comment]
ok thanks, I finish the tests

**Comment by vincentbz** [comment]
OK so tests are good :)
Voucher applied when product is not discounted. If only a combination of the product is discounted, the voucher is applied on other combinations.
If I add a product discounted AND a product without discount, the voucher (ex 10%) is applied only on the "normal" product

**Comment by kpodemski** [comment]
Nice, looking forward for merge, i'll work on amount reduction next week

File: classes/CartRule.php [hunk 1]
```diff
@@ -578,6 +580,20 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
...
+            $products = $context->cart->getProducts();
+            $is_ok = false;
+            foreach ($products as $product) {
+                if (!$product['reduction_applies']) {
+                    $is_ok = true;
+                    break;
+                }
+            }
+            if (!$is_ok) {
+                return (!$display_error) ? false : Tools::displayError('You cannot use this voucher on reduced products');
```
Comment (mickaelandrieu) [inline classes/CartRule.php]: shouldn't be translated ?

Comment (kpodemski) [inline classes/CartRule.php]: I don't think so, unless something changed, it is available to translate from Localization -> Translations -> Error messages, look on other places in this file:
https://github.com/kpodemski/PrestaShop/blob/3b1a3535dd1ddfce860eb27fecfb8e65a22b5c9c/classes/CartRule.php#L514

Comment (mickaelandrieu) [inline classes/CartRule.php]: agree :+1:

**Comment by mickaelandrieu** [comment]
I merge, thank you @kpodemski

**Comment by PrestaShark** [comment]
@kpodemski and @djbuch well done! This is most wanted feature for many years! 
Sorry for trouble but it might be considered too:
- Disable voucher for products with quantity discounts
- Disable voucher for strict user group - a new one like B2B, Wholesale to not be able to use their discounts with well known and existing vouchcer codes like this for Newsletter or Facebook etc.

For now this new wholesale user group can use vouchers without any problem and its totaly unacceptable! Consider this feature to!

Thanks!
Konrad

**Comment by AlexEven** [comment]
@vincentbz + @xBorderie and @Shudrum  -- please do not forget about the wording label!
I discovered this feature as I'm updating the documentation and it isn't clear at all!
I'll rewrite the feature's label + add a hint to explain what it's all about.
Thank you!

**Comment by AlexEven** [comment]
And we should avoid negative formulation - why is it "exclude on specials" disabled (so it's applying on specials in the end?) and not something like "Apply on discounted products" disabled by default?

**Pull request merged**